---
output: html_document
title: "UKB exploration of genome-wide GxMedDiet interactions"
---


```{r setup, include=F}
knitr::opts_chunk$set(echo=F, message=F,warning=F, dev="png", dpi=300,
                      fig.path="../output/gwis_summary/",
                      cache.path="../cache/gwis_summary/")
suppressMessages(silent <- lapply(
  c("knitr", "kableExtra", "data.table", "tidyverse", "patchwork"), 
  library, character.only=T))
```

# Genetic correlations and the inflammation-glycemia relationship

```{r genetic-correlations}
glycemic_inflammatory_biomarkers <- c("hba1c", "glu", "hscrp_log", "aisi_log", "sii_log", "siri_log")

read_rg <- function(f) {
  log_obj <- readLines(f)
  rg_str <- grep("Genetic Correlation: ", log_obj, value=TRUE)
  rg <- as.numeric(str_split(rg_str, " ")[[1]][3])
  p_str <- grep("P: ", log_obj, value=TRUE)
  p <- as.numeric(str_split(p_str, " ")[[1]][2])
  tibble(
    rg = rg,
    p = p
  )
}

rg_df <- expand_grid(
  bm1 = glycemic_inflammatory_biomarkers,
  bm2 = glycemic_inflammatory_biomarkers
) %>%
  rowwise() %>%
  mutate(rg_res = read_rg(paste0("../data/processed/ldsc/", bm1, "_", bm2, ".log"))) %>%
  unnest(rg_res)

rg_df %>%
  mutate(across(rg, round, 2)) %>%
  ggplot(aes(x=bm1, y=bm2, fill=rg)) +
  geom_tile() +
  geom_text(aes(label=rg), color="white", size=4) +
  coord_fixed() +
  labs(title="Genetic correlations")
```

```{r prs}
extract_prs_lm <- function(f) {
  readRDS(f) %>%
    broom::tidy() %>%
    filter(term == "score")
}

prs_df <- expand_grid(
  bm1 = c("hba1c", "glu", "hscrp_log", "aisi_log", "sii_log", "siri_log"),
  bm2 = c("hba1c", "glu", "hscrp_log", "aisi_log", "sii_log", "siri_log")
) %>%
  rowwise() %>%
  mutate(prs_res = extract_prs_lm(paste0("../data/processed/prs/", bm1, "_prs_", bm2, "_lm_fit.rds"))) %>%
  unnest(prs_res)

prs_df %>%
  mutate(across(statistic, round, 1)) %>%
  ggplot(aes(x=bm1, y=bm2, fill=statistic)) +
  geom_tile() +
  geom_text(aes(label=statistic), color="white") +
  coord_fixed() +
  labs(title="PRS effects (BM1 -> BM2)")
```

```{r md-hscrp, fig.width=3, fig.asp=1, eval=F}
lm_str <- paste0("hscrp_log ~ age + sex + ", paste0("gPC", 1:5, collapse=" + "))
phenos$hscrp_log_resid <- resid(lm(as.formula(lm_str), data=phenos, na.action=na.exclude))

# phenos %>%
#   ggplot(aes(x=med_score, y=hscrp_log)) +
#   geom_smooth()

mds_crp_resid_boxplot_plt <- phenos %>%
  filter(!is.na(med_score)) %>%
  ggplot(aes(x=factor(med_score), y=hscrp_log_resid)) +
  geom_boxplot() +
  labs(x="MDS",
       y="log(hsCRP) residuals")

mds_crp_resid_smooth_plt <- phenos %>%
  ggplot(aes(x=med_score, y=hscrp_log_resid)) +
  geom_smooth() +
  labs(x="MDS",
       y="log(hsCRP) residuals")

mds_hist_plt <- phenos %>%
  ggplot(aes(x=med_score)) +
  geom_histogram() +
  labs(x="MDS")

mds_crp_resid_plt / mds_hist_plt

plot_grid(mds_crp_resid_boxplot_plt, mds_hist_plt, 
          ncol=1, rel_heights=c(2, 1),
          align="hv")
```

Genome-wide interaction study model:

$biomarker ~ G + MedDiet + G*MedDiet + age + sex + age*sex + 5 gPCs + 5 gPC*MedDiet$

Here, we're using the genome-wide approach for two reasons:

1. To establish the validity of the model (via Q-Q plots and genomic inflation).
2. To prioritize candidate genes for more in-depth analysis, both in UKB (for example, using rare variants) and later in TOPMed.

# Genome-wide summaries

```{r model-data}
biomarkers <- c("hscrp_log", "aisi_log", "hba1c")
# biomarkers <- c("hscrp_log", "hba1c")
models <- c("MD", "MDcomponents")
mds_components <- scan("../data/processed/meddiet_components.txt", what=character())
mds_components_ext <- c(mds_components, 
                        "NUTSbin", "LEGUMESbin", "FISHbin", "fish_oil", "oily_fish", "nonoily_fish")
diet_vars <- c("WHGRAIN", "FRUIT", "FISH", "NUTS", "NUTSbin", "LEGUMESbin", "FISHbin", "fish_oil", "oily_fish", "nonoily_fish")
```

```{r summary-functions}
calc_lambda <- function(x, p=0.5){
  # Calculate genomic inflation lambda value
  x <- x[!is.na(x)]
  x.quantile <- quantile(x, p)
  round(qchisq(1 - x.quantile, 1) / qchisq(p, 1), 2)
}

make_qq <- function(data, pval_col, main=""){
  # Make a quantile-quantile plot
  data <- filter(data, data[[pval_col]] > 0)  # In case extremely low p-values are stored as zero

  # Process p-values
  y_vals <- sort(-log10(data[[pval_col]]))
  x_vals <- -log10(rev(ppoints(length(y_vals))))  # ppoints generates a uniform probability distribution
  lam <- calc_lambda(data[[pval_col]])

  # Trim points at higher p-values (credit to RaMWAS package for code snippet)
  levels = as.integer((x_vals - x_vals[1]) / (tail(x_vals, 1) - x_vals[1]) * 2000)
  keep = c(TRUE, diff(levels) != 0)
  levels = as.integer((y_vals - y_vals[1])/(tail(y_vals, 1) - y_vals[1]) * 2000)
  keep = keep | c(TRUE, diff(levels) != 0)
  keep = which(keep)

  rm(data)  # Deal with memory issues
    
  tibble(x=x_vals[keep], y=y_vals[keep]) %>%
    ggplot(aes(x=x, y=y)) + 
    geom_point() +
    geom_abline(slope=1, linetype="dashed") +
    annotate("text", 
#              label=bquote(lambda == .(lam)),  # NOTE: somehow this bquote call results in the warning about "is.na() applied to type language"
             label=paste("lambda == ", lam), parse=TRUE,
             x=-Inf, y=Inf, hjust=-0.5, vjust=3, size=5) +
    labs(x=expression(-log[10](italic(p)) * " (Expected)"),
         y=expression(-log[10](italic(p)) * " (Observed)"),
         title=main) +
    theme_bw() +
    theme(panel.grid=element_blank())
}

make_manhattan <- function(data, pval_col, snp_col="SNP", chr_col="CHR", pos_col="POS", 
                           threshold=5e-8, ylims=NULL, main="") {

  mh_data <- data %>%
    dplyr::rename(all_of(c(SNP=snp_col, CHR=chr_col, POS=pos_col, P=pval_col))) %>%
    filter(!is.na(P)) %>%
    mutate(P = as.numeric(P),
           P = ifelse(P == 0, min(1e-300, min(P[P != 0], na.rm=T)), P),  # Remove P = 0
           nlp = -log10(P))
  
  # Trim points in crowded regions (credit to RaMWAS package for code snippet)
  yfac = as.integer(mh_data$nlp * 100) + 1L
  yorder = sort.list(yfac)
  yfac <- factor(yfac, levels=as.character(seq_len(max(yfac))))
  ygroup <- split(seq_along(yfac), yfac)
  for (i in seq_along(ygroup)) {
    if (length(ygroup[[i]]) > 300) {
      ygroup[[i]] <- sample(ygroup[[i]], size=300, replace=F)
    }
  }
  keep <- unlist(ygroup, use.names=FALSE)
  
  mh_data <- mh_data %>%
    select(SNP, CHR, POS, nlp) %>%
    dplyr::slice(keep) %>%
    mutate(POS = as.numeric(as.character(POS)),
           CHR = factor(CHR, levels=c(1:22, "X"))) %>%
    arrange(CHR, POS) %>%
    mutate(pos_idx = seq(1, nrow(.)))
  
  suppressWarnings(chr_lengths <- sapply(c(1:22, "X"), function(chr) with(mh_data, max(POS[CHR == chr], na.rm=T))))
  chr_lengths <- ifelse(is.infinite(chr_lengths), 0, chr_lengths)
  chr_start_pos <- cumsum(chr_lengths) - chr_lengths
  
  mh_data <- mh_data %>%
    mutate(x_coord = chr_start_pos[CHR] + POS,
           color = CHR,
    ) %>%
    arrange(as.integer(color), nlp) #%>%
#     distinct(SNP, .keep_all=T)
  
  lims <- mh_data %>%
    group_by(CHR) %>%
    summarise(avg_coord = (min(x_coord) + max(x_coord)) / 2)
  
  newcols <- setNames(
    rep(x=c("#AAAAAA", "#8A8A8A"), length.out=23),  # Gray/dark gray for alternating chromosomes
    levels(factor(lims$CHR))
  )
  
  p1 <- ggplot() +
    geom_point(data=mh_data, 
               aes(x=x_coord, y=nlp, color=factor(color)), 
               size=0.75, alpha=1) +
    geom_hline(yintercept=-log10(threshold), linetype="dashed", color="black") + 
    scale_x_continuous(breaks=lims$avg_coord[c(1:16, 18, 20, 20, 22, 23)], 
                       labels=c(1:16, 18, 20, 20, 22, "X"), 
                       expand=c(0,0)) +
    scale_y_continuous(name=expression(-log[10](italic(p)))) +
    scale_colour_manual(values=newcols, guide="none") +
    scale_fill_manual(name="Color", values=newcols) +
    labs(title=main) +
    theme(axis.title.x=element_blank(),
          axis.text.x=element_text(vjust = -1.5),
          panel.grid.minor=element_blank(),
          panel.grid.major.x=element_blank())
  if (!is.null(ylims)) p1 <- p1 + coord_cartesian(ylim=ylims)
  
  p1
}
```

## Whole-genome

```{r gw-gxe}
read_gw_gxe <- function(bm, dv) {
  f <- paste0("../data/processed/gw_gxe/", bm, "_", dv, "_pigeon_h2_GxE_Var.txt")
  if (!file.exists(f)) return(tibble())
  read_table(f)
}

read_gw_gxe_ldsc <- function(bm, dv) {
  f <- paste0("../data/processed/gw_gxe/", bm, "_", dv, "_ldsc_h2.log")
  if (!file.exists(f)) return(tibble())
  h2_line <- grep("Total Observed scale h2:", readLines(f), value=TRUE)
  res_vec <- str_split(h2_line, " ")[[1]]
  print(gsub("\\(|\\)", "", res_vec[6]))
  tibble(
    bm = bm,
    dv = dv,
    h2_gxe_obs = as.numeric(res_vec[5]),
    h2_gxe_obs_se = as.numeric(gsub("\\(|\\)", "", res_vec[6]))
  ) %>%
    mutate(h2_gxe_obs_p = pchisq((h2_gxe_obs / h2_gxe_obs_se)^2, df=1, lower.tail=FALSE))
}

expand_grid(
  bm = c("hscrp_log", "aisi_log", "hba1c"),
  dv = mds_components_ext
) %>%
  rowwise() %>%
  mutate(pigeon_res = list(read_gw_gxe(bm, dv)),
         ldsc_res = list(read_gw_gxe(bm, dv))) %>%
  unnest(ldsc_res)
```

## Variant-level

```{r load-variant-sumstats}
read_nom_sumstats <- function(bm, model) {
  fn <- paste0("../data/processed/gwis/", bm, "_", model, "_gwis_merged_nom")
  read_tsv(fn) %>%
    select(SNP, CHR, POS, EA, NEA, AF, P)
}

gwis_res_list <- lapply(setNames(biomarkers, biomarkers), function(bm) {
  lapply(setNames(models, models), function(model) {
    read_nom_sumstats(bm, model)
  })
})
```

```{r variant-summaries}
for (bm in biomarkers) {
  for (dv in diet_vars) {
    f <- paste0("../data/processed/gwis/qq_plots/", bm, "_", dv, "_gwis_QQ.pdf")
    if (file.exists(f)) include_graphics(f)
  }
}

# make_manhattan(gwis_res_list$hscrp_log$MD, "P")
# make_manhattan(gwis_res_list$hscrp_log$MDcomponents, "P")
# make_manhattan(gwis_res_list$hba1c$MD, "P")
# make_manhattan(gwis_res_list$hba1c$MDcomponents, "P")
```

## Gene-level

Using the FUMA GWAS platform, MAGMA was run to generate gene-level interaction test p-values.

```{r load-gene-sumstats}
read_gene_sumstats <- function(bm, model) {
  fn <- paste0("../data/processed/magma/", bm, "_", model, ".genes.out")
  if (!file.exists(fn)) return(tibble())
  read_tsv(fn, col_types=cols()) %>%
    select(ens_gene=GENE, gene=SYMBOL, z=ZSTAT, P) %>%
    mutate(q = p.adjust(P, method="BH"))
}

gene_res_df <- expand_grid(
  biomarker = biomarkers,
  diet_var = diet_vars
) %>%
  rowwise() %>%
  mutate(gene_res = list(read_gene_sumstats(biomarker, diet_var))) %>%
  unnest(gene_res)
```

```{r gene-summaries}
for (bm in biomarkers) {
  for (dv in diet_vars) {
    gene_res <- filter(gene_res_df, biomarker == bm, diet_var == dv)
    if (nrow(gene_res) > 0) {
      qq_plt <- make_qq(gene_res, "P", paste0(bm, " - ", dv))
      print(qq_plt)
      top_res <- head(arrange(gene_res, P))
      print(top_res)
    } 
  }
}
```

```{r fish-comparisons}
fish_vars <- c("FISH", "FISHbin", "fish_oil", "oily_fish", "nonoily_fish")
fish_gene_res_df <- gene_res_df %>%
  # filter(biomarker == "hscrp_log") %>%
  filter(diet_var %in% fish_vars) %>%
  select(biomarker, diet_var, gene, z) %>%
  # mutate(P = -log10(P)) %>%
  pivot_wider(names_from="diet_var", values_from="z") %>%
  mutate(avg = (FISHbin + fish_oil + oily_fish + nonoily_fish) / 4)

fish_comparison_plots <- function(v1, v2, bm) {
  fish_gene_res_df %>%
    filter(biomarker == bm) %>%
    ggplot(aes_string(x=v1, y=v2)) +
    geom_point() +
    geom_abline(color="gray")
}

fish_comparison_plots("FISHbin", "fish_oil", "hscrp_log")
fish_comparison_plots("FISHbin", "fish_oil", "aisi_log")

fish_gene_res_df %>%
  filter(biomarker == "hscrp_log") %>%
  select(gene, FISHbin:nonoily_fish) %>%
  pivot_longer(FISHbin:nonoily_fish, names_to="diet_var", values_to="z") %>%
  head()

fish_gene_cor_res_df <- expand_grid(
  dv1 = fish_vars,
  dv2 = fish_vars
) %>%
  rowwise() %>%
  mutate(cor_res = list(broom::tidy(cor.test(fish_gene_res_df[[dv1]], fish_gene_res_df[[dv2]])))) %>%
  unnest(cor_res)
fish_gene_cor_res_df %>%
  mutate(estimate = round(estimate, 3)) %>%
  ggplot(aes(x=dv1, y=dv2, fill=estimate)) +
  geom_tile() +
  geom_label(aes(label=estimate))
```

```{r asn-abstract-results}
bonferroni <- 0.05 / nrow(filter(gene_res_df, diet_var == "FISHbin", biomarker == "hscrp_log"))

top_gene_res_df <- gene_res_df %>%
  # filter(biomarker == "hscrp_log") %>%
  filter(diet_var %in% c(fish_vars, "NUTSbin", "FRUIT", "WHGRAIN", "NUTS")) %>%
  select(biomarker, diet_var, gene, P) %>%
  mutate(P = -log10(P)) %>%
  pivot_wider(names_from="diet_var", values_from="P") %>%
  mutate(avg = (FISHbin + NUTSbin + FRUIT + WHGRAIN) / 4)

top_gene_res_df %>%
  filter(biomarker == "hscrp_log") %>%
  select(fish_vars, NUTSbin, FRUIT, WHGRAIN, NUTS) %>%
  mutate(across(everything(), ~. > -log10(bonferroni))) %>%
  colSums()

bonferroni_gene_res_df <- top_gene_res_df %>%
  filter_at(vars(-biomarker, -gene), any_vars(. > -log10(bonferroni)))
bonferroni_gene_res_df

gene_res_df %>%
  filter(diet_var %in% c("FISHbin", "NUTSbin"),
         gene %in% bonferroni_gene_res_df$gene) %>%
  select(biomarker, diet_var, gene, z) %>%
  pivot_wider(names_from="biomarker", values_from="z") %>%
  ggplot(aes(x=hscrp_log, y=aisi_log, color=diet_var)) +
  geom_point() +
  geom_abline(linetype="dashed") +
  geom_label(aes(label=gene))
```

Is there a major difference between the gene-level interaction p-values from the single MDS exposure model versus the 9-exposure components model?

```{r compare-md-and-components, fig.width=3, eval=F}
# model_comparison_df <- bind_rows(lapply(gene_res_list, bind_rows, .id="model"), 
#                                  .id="biomarker") %>%
model_comparison_df <- gene_res_df %>%
  
  mutate(nlp = -log10(P)) %>%
  select(biomarker, model, ens_gene, nlp) %>%
  pivot_wider(names_from="diet_var", values_from="nlp") %>%
  filter(biomarker == "hscrp_log") 

model_comparison_df %>%
  ggplot(aes(x=MD, y=MDcomponents)) +
  geom_point() +
  geom_abline(linetype="dashed")

cor.test(~ MD + MDcomponents, data=model_comparison_df)
```

## Read in auxiliary data
 
### Gene-level evidence

* Main effect gene burden tests from ExPheWAS
* GWS main effects from vQTL paper (w/ gene annotations)
* GWS vQTL effects from vQTL paper (w/ gene annotations)

```{r exphewas}
bm_exphewas_name_map <- c(
  hba1c="Glycated haemoglobin (HbA1c)", glu="Glucose", apoB="Apolipoprotein B", 
  ldl="LDL direct", chol="Cholesterol", hdl="HDL cholesterol",
  tg_log="Triglycerides", hscrp_log="C-reactive protein", 
  alt_log="Alanine aminotransferase", vitD="Vitamin D"
)

exphewas_df <- lapply(bm_exphewas_name_map, function(bm) {
  fn <- grep(bm, list.files("../data/raw/exphewas/", full.names=TRUE), value=TRUE)
  read_csv(fn)
}) %>%
  setNames(names(bm_exphewas_name_map)) %>%
  bind_rows(.id="biomarker")  %>%
  mutate(sig_exphewas = `Q-Value` < 0.05) %>%
  select(biomarker, gene=Symbol, sig_exphewas, `Q-Value`)


slc_variant_anno <- read_tsv("../data/processed/slc_gene_variant_list.tsv") %>%
  select(rsid=`Variant name`, gene=`Gene name`, consequence=`Variant consequence`)

slc_exphewas_df <- exphewas_df %>%
  filter(gene %in% slc_variant_anno$gene)
```

```{r vqtl-data}
bm_vqtl_name_map <- c(
  hba1c="hba1c_log", glu="glu_log", apoB="apoB_statinadj_log", 
  ldl="ldl_statinadj_log", chol="chol_statinadj_log", hdl="hdl_log",
  tg_log="tg_log", hscrp_log="hscrp_log", 
  alt_log="alt_log"#, vitD="vitD"
)

vqtl_study_df <- lapply(bm_vqtl_name_map, function(bm) {
  vf_colnames <- c("type", "gene", "CHR", "POS", "end", "a1", "a2", "rsID", "P")
  me_vf_df <- read_tsv(paste0("../data/processed/prior_evidence/", bm, "_me.variant_function"),
                       col_names=vf_colnames)
  vqtl_vf_df <- read_tsv(paste0("../data/processed/prior_evidence/", bm, "_vqtl.variant_function"),
                         col_names=vf_colnames)
  vf_df <- bind_rows(list(me=me_vf_df, vqtl=vqtl_vf_df), .id="type") %>%
    filter(P < 5e-8) %>%
    select(type, CHR, POS, gene) %>%
    mutate(CHR = as.character(CHR)) %>% 
    separate_rows(gene, sep=",") %>% 
    mutate(gene_dist = gsub("dist=", "", str_extract(gene, "dist=[0-9]*")),
           gene_dist = as.numeric(gene_dist),
           gene = gsub("\\(.*", "", gene)) %>%
    # group_by(CHR, POS) %>%
    # arrange(gene_dist) %>%
    # dplyr::slice(1) %>%
    # ungroup() %>%
    # select(CHR, POS, gene)
    # filter(grepl("^SLC", gene)) %>%
    distinct(type, gene) %>%
    mutate(value=TRUE) %>%
    pivot_wider(names_from="type", names_prefix="sig_",
                values_from="value", values_fill=FALSE)
}) %>%
  bind_rows(.id="biomarker")
```

```{r gene-evidence, fig.width=5}
gene_evidence_df <- exphewas_df %>%
  full_join(vqtl_study_df, by=c("gene", "biomarker")) %>%
  mutate(across(contains("sig_"), ~ifelse(is.na(.), FALSE, .)))

gene_evidence_df %>%
  group_by(biomarker) %>%
  summarise(across(contains("sig_"), sum)) %>%
  kable()

gene_annot_list <- lapply(setNames(biomarkers, biomarkers), function(bm) {
  df <- filter(gene_evidence_df, biomarker == bm)
  list(
    ExPheWAS = df$gene[df$sig_exphewas],
    ME = df$gene[df$sig_me],
    vQTL = df$gene[df$sig_vqtl],
    SLC = grep("^SLC", df$gene, value=TRUE),
    ME_and_vQTL = df$gene[df$sig_me & df$sig_vqtl]
  )
})

ggVennDiagram::ggVennDiagram(gene_annot_list$hscrp_log)
ggVennDiagram::ggVennDiagram(gene_annot_list$hba1c)
```

```{r permutation-enrichment-tests}
run_perm_test <- function(p_test, p_all, n_bg, random_seed=123) {
  set.seed(random_seed)
  chisq_test <- qchisq(p_test, df=1, lower.tail=FALSE)
  chisq_test_sum <- sum(chisq_test)
  chisq_all <- qchisq(p_all, df=1, lower.tail=FALSE)
  chisq_bg_sums <- sapply(1:n_bg, function(i) {
    sum(sample(chisq_all, length(chisq_test), replace=FALSE))
  })
  p_out <- sum(chisq_bg_sums > chisq_test_sum) / n_bg
  p_out
}
```

```{r gene-category-enrichments}
gene_cat_enrich_df <- expand_grid(
  biomarker = biomarkers,
  model = models,
  gene_set = names(gene_annot_list$hscrp_log)
) %>%
  mutate(perm_p = pmap_dbl(list(biomarker, model, gene_set), function(b, m, g) {
    df <- gene_res_list[[b]][[m]]
    run_perm_test(filter(df, gene %in% gene_annot_list[[b]][[g]])$P, df$P, 1e4)
  }))

gene_cat_enrich_df

#   tibble(
#   gene_category = c("SLC genes", "ExPheWAS main effects", "Main effects from Westerman 2022", "vQTLs from Westerman 2022"),
#   `Enrichment P` = c(p_slc_enrich, p_exphewas_enrich, p_me_enrich, p_vqtl_enrich)
# ) %>%
#   kable(caption="Enrichment of gene-based GWIS signal in various gene categories")
```

```{r enrichments-tmp-hba1c}
ggVennDiagram::ggVennDiagram(gene_annot_list$hba1c[c("ME", "vQTL")],
                             label="count") +
  theme(legend.position="none")

gene_cat_enrich_df %>%
  filter(biomarker == "hba1c",
         model == "MDcomponents",
         gene_set %in% c("ME", "vQTL", "ME_and_vQTL", "ExPheWAS")) %>%
  select(`Gene set`=gene_set, `Permutation P-value`=perm_p)

make_qq(filter(gene_res_list$hba1c$MDcomponents, 
               gene %in% gene_annot_list$hba1c$ME_and_vQTL),
        "P")

filter(gene_res_list$hba1c$MDcomponents, 
               gene %in% gene_annot_list$hba1c$ME_and_vQTL) %>%
  arrange(P) %>%
  mutate(q = p.adjust(P, method="BH")) %>%
  head()
```

```{r enrichments-tmp-hscrp}
ggVennDiagram::ggVennDiagram(gene_annot_list$hscrp_log[c("ME", "vQTL")],
                             label="count") +
  theme(legend.position="none")

gene_cat_enrich_df %>%
  filter(biomarker == "hscrp_log",
         model == "MDcomponents",
         gene_set %in% c("ME", "vQTL", "ME_and_vQTL")) %>%
  select(`Gene set`=gene_set, `Permutation P-value`=perm_p)

make_qq(filter(gene_res_list$hscrp_log$MDcomponents, 
               gene %in% gene_annot_list$hscrp_log$ME),
        "P")

filter(gene_res_list$hscrp_log$MDcomponents, 
               gene %in% gene_annot_list$hscrp_log$ME) %>%
  arrange(P) %>%
  mutate(q = p.adjust(P, method="BH")) %>%
  head()
```

# Gene-specific follow-up

```{r bhmt-sumstats}
bhmt_ss <- read_tsv("../data/processed/gwis/hscrp_log_MDcomponents_gwis_bhmt_subset")
for (dv in mds_components) {
  beta_col <- paste0("Beta_G-", dv)
  var_col <- paste0("Var_Beta_G-", dv)
  chisq_vals <- bhmt_ss[[beta_col]]^2 / bhmt_ss[[var_col]]
  bhmt_ss[[paste0("p_int_", dv)]] <- pchisq(chisq_vals, df=1, lower.tail=FALSE)
}

bhmt_ss %>%
  select(POS, contains("p_int")) %>%
  pivot_longer(-POS, names_to="dv", names_prefix="p_int_", values_to="p_int") %>%
  mutate(nlp_int = -log10(p_int)) %>% View()
  ggplot(aes(x=POS, y=nlp_int, color=dv)) +
  geom_point() +
  scale_color_discrete(name="MDS component") +
  labs(y=expression(Exposure-specific-log[10](p)))
```

```{r bhmt-regressions}
# library(vcfR)
# a <- read.vcfR("../data/processed/genotypes/bhmt_genotypes.vcf.gz")

gt_mat <- a@gt
var_info <- a@fix

phenos <- read_csv("../data/processed/ukb_phenos_unrelated.csv", col_types=cols(id="c"))
rs622087_vec <- gt_mat[which(var_info[, "ID"] == "rs622087"), ]
rs622087_df <- tibble(id = names(rs622087_vec)[-1], rs622087 = rs622087_vec[-1]) %>%
  mutate(rs622087 =  as.numeric(gsub(".*:", "", rs622087)))
phenos <- left_join(phenos, rs622087_df, by="id")

covars <- c("age", "sex", "ageBysex", 
            paste0("gPC", 1:5), 
            paste0("med_scoreBygPC", 1:5))
lm_str <- paste0("hscrp_log ~ rs622087 * FISH + ", paste(covars, collapse=" + "))
lm_fit <- lm(as.formula(lm_str), data=phenos)
broom::tidy(lm_fit)

lm_str <- paste0("hscrp_log ~ rs622087 * oily_fish + ", paste(covars, collapse=" + "))
lm_fit <- lm(as.formula(lm_str), data=phenos)
broom::tidy(lm_fit)
lm_str <- paste0("hscrp_log ~ rs622087 * nonoily_fish + ", paste(covars, collapse=" + "))
lm_fit <- lm(as.formula(lm_str), data=phenos)
broom::tidy(lm_fit)

lm_str <- paste0("hscrp_log ~ rs622087 * fish_oil + ", paste(covars, collapse=" + "))
lm_fit <- lm(as.formula(lm_str), data=phenos)
broom::tidy(lm_fit)

phenos %>%
  filter(!is.na(FISH), !is.na(hscrp_log)) %>%
  mutate(fish_tert = cut(FISH, c(-1, 0, 1, 10), include.lowest=FALSE,
                         labels=c("None", "<1 serving/day", ">=1 serving/day")),
         rs622087_int = factor(round(rs622087), labels=c("TT", "GT", "GG"))) %>%
  group_by(fish_tert, rs622087_int) %>%
  summarise(m = mean(hscrp_log),
            se = sd(hscrp_log) / sqrt(n()),
            l95 = m - 1.96 * se,
            u95 = m + 1.96 * se) %>%
  ggplot(aes(x=rs622087_int, y=m, color=fish_tert)) +
  geom_point(position=position_dodge(width=0.2)) +
  geom_errorbar(aes(ymin=l95, ymax=u95), position=position_dodge(width=0.2),
                width=0.2) +
  scale_color_discrete(name="Fish intake (24HR)") +
  labs(x="rs622087",
       y="Mean log(hsCRP) w/ 95% CI")

phenos %>%
  filter(!is.na(oily_fish), !is.na(hscrp_log)) %>%
  mutate(fish_tert = cut(oily_fish, c(-1, 0, 0.2, 10), include.lowest=FALSE,
                         labels=c("None", "<=1 serving/wk", ">1 serving/wk")),
         rs622087_int = factor(round(rs622087), labels=c("TT", "GT", "GG"))) %>%
  group_by(fish_tert, rs622087_int) %>%
  summarise(m = mean(hscrp_log),
            se = sd(hscrp_log) / sqrt(n()),
            l95 = m - 1.96 * se,
            u95 = m + 1.96 * se) %>%
  ggplot(aes(x=rs622087_int, y=m, color=fish_tert)) +
  geom_point(position=position_dodge(width=0.2)) +
  geom_errorbar(aes(ymin=l95, ymax=u95), position=position_dodge(width=0.2),
                width=0.2) +
  scale_color_discrete(name="Oily fish intake (FFQ)") +
  labs(x="rs622087",
       y="Mean log(hsCRP) w/ 95% CI")

phenos %>%
  filter(!is.na(nonoily_fish), !is.na(hscrp_log)) %>%
  mutate(fish_tert = cut(nonoily_fish, c(-1, 0, 0.2, 10), include.lowest=FALSE,
                         labels=c("None", "<=1 serving/wk", ">1 serving/wk")),
         rs622087_int = factor(round(rs622087), labels=c("TT", "GT", "GG"))) %>%
  group_by(fish_tert, rs622087_int) %>%
  summarise(m = mean(hscrp_log),
            se = sd(hscrp_log) / sqrt(n()),
            l95 = m - 1.96 * se,
            u95 = m + 1.96 * se) %>%
  ggplot(aes(x=rs622087_int, y=m, color=fish_tert)) +
  geom_point(position=position_dodge(width=0.2)) +
  geom_errorbar(aes(ymin=l95, ymax=u95), position=position_dodge(width=0.2),
                width=0.2) +
  scale_color_discrete(name="Non-oily fish intake (FFQ)") +
  labs(x="rs622087",
       y="Mean log(hsCRP) w/ 95% CI")

phenos %>%
  filter(!is.na(fish_oil), !is.na(hscrp_log)) %>%
  mutate(fish_tert = cut(fish_oil, c(-1, 0, 1), include.lowest=FALSE,
                         labels=c("No", "Yes")),
         rs622087_int = factor(round(rs622087), labels=c("TT", "GT", "GG"))) %>%
  group_by(fish_tert, rs622087_int) %>%
  summarise(m = mean(hscrp_log),
            se = sd(hscrp_log) / sqrt(n()),
            l95 = m - 1.96 * se,
            u95 = m + 1.96 * se) %>%
  ggplot(aes(x=rs622087_int, y=m, color=fish_tert)) +
  geom_point(position=position_dodge(width=0.2)) +
  geom_errorbar(aes(ymin=l95, ymax=u95), position=position_dodge(width=0.2),
                width=0.2) +
  scale_color_discrete(name="Fish oil intake (24HR)") +
  labs(x="rs622087",
       y="Mean log(hsCRP) w/ 95% CI")
```

```{r hnf1a}
hnf1a_ss <- read_tsv("../data/processed/gwis/hscrp_log_MDcomponents_gwis_hnf1a_subset")
for (dv in mds_components) {
  beta_col <- paste0("Beta_G-", dv)
  var_col <- paste0("Var_Beta_G-", dv)
  chisq_vals <- hnf1a_ss[[beta_col]]^2 / hnf1a_ss[[var_col]]
  hnf1a_ss[[paste0("p_int_", dv)]] <- pchisq(chisq_vals, df=1, lower.tail=FALSE)
}

hnf1a_ss %>%
  select(POS, contains("p_int")) %>%
  pivot_longer(-POS, names_to="dv", names_prefix="p_int_", values_to="p_int") %>%
  mutate(nlp_int = -log10(p_int)) %>%
  ggplot(aes(x=POS, y=nlp_int, color=dv)) +
  geom_point()
```

```{r mapk3}
mapk3_ss <- read_tsv("../data/processed/gwis/hscrp_log_MDcomponents_gwis_mapk3_subset")
for (dv in mds_components) {
  beta_col <- paste0("Beta_G-", dv)
  var_col <- paste0("Var_Beta_G-", dv)
  chisq_vals <- mapk3_ss[[beta_col]]^2 / mapk3_ss[[var_col]]
  mapk3_ss[[paste0("p_int_", dv)]] <- pchisq(chisq_vals, df=1, lower.tail=FALSE)
}

mapk3_ss %>%
  select(POS, contains("p_int")) %>%
  pivot_longer(-POS, names_to="dv", names_prefix="p_int_", values_to="p_int") %>%
  mutate(nlp_int = -log10(p_int)) %>%
  ggplot(aes(x=POS, y=nlp_int, color=dv)) +
  geom_point()
```

```{r read-vcf}
library(vcfR)
a <- read.vcfR("../data/processed/genotypes/mapk3_genotypes.vcf.gz")

gt_mat <- a@gt
gt_mat <- gsub(".*:", "", gt_mat)
```



```{r load-data, eval=F}
read_bm_res <- function(bm) {
  read_tsv(paste0("../data/processed/ukb_gxe_res/", bm, "_minimal_slc")) %>%
    mutate(CHR = gsub("^0", "", CHR))
}

res_list <- lapply(setNames(biomarkers, biomarkers), function(bm) {
  read_bm_res(bm)
})
combined_res <- bind_rows(res_list, .id="biomarker")
```

```{r primary-results, eval=F}
primary_res <- combined_res %>%
  left_join(variant_anno, by=c("RSID"="rsid")) %>%
  left_join(gene_evidence_df, by=c("biomarker", "gene"))
```

```{r manhattans, eval=F}
mh_data <- res_list$hba1c  %>%
  select(SNP=RSID, CHR, POS, P=P_Value_Interaction) %>%
  filter(!is.na(P)) %>%
  mutate(CHR = as.integer(ifelse(CHR == "X", 23, CHR)),
         P = ifelse(P == 0, min(1e-300, min(P[P != 0], na.rm=T)), P),  # Remove P = 0
         nlp = -log10(P))

# Trim points in crowded regions (credit to RaMWAS package for code snippet)
yfac = as.integer(mh_data$nlp * 100) + 1L
yorder = sort.list(yfac)
yfac <- factor(yfac, levels=as.character(seq_len(max(yfac))))
ygroup <- split(seq_along(yfac), yfac)
for (i in seq_along(ygroup)) {
  if (length(ygroup[[i]]) > 300) {
    ygroup[[i]] <- sample(ygroup[[i]], size=300, replace=F)
  }
}
keep <- unlist(ygroup, use.names=FALSE)

mh_data <- mh_data %>%
  select(SNP, CHR, POS, nlp) %>%
  dplyr::slice(keep) %>%
  mutate(POS = as.numeric(as.character(POS)),
         CHR = factor(CHR, levels=1:22)) %>%
  arrange(CHR, POS) %>%
  mutate(pos_idx = seq(1, nrow(.)))

chr_lengths <- sapply(1:23, function(chr) {
  if (chr %in% mh_data$CHR) max(mh_data$POS[mh_data$CHR == chr]) else 1e8
} )
chr_start_pos <- cumsum(chr_lengths) - chr_lengths

mh_data <- mh_data %>%
  mutate(x_coord = chr_start_pos[CHR] + POS,
         color = CHR,
  ) %>%
  arrange(as.integer(color), nlp)

lims <- mh_data %>%
  group_by(CHR) %>%
  summarise(avg_coord = (min(x_coord) + max(x_coord)) / 2)

mh_cols <- setNames(
  rep(x=c("#999999", "#555555"), length.out=22),  # Gray/dark gray for alternating chromosomes
  levels(factor(lims$CHR))
)

ggplot() +
  geom_point(data=mh_data, 
             aes(x=x_coord, y=nlp, color=factor(color)), 
             size=0.6, alpha=1) +
  # geom_hline(yintercept=-log10(bonferroni), linetype="dashed", color="black") + 
  scale_x_continuous(breaks=lims$avg_coord[c(1:16, 18, 20, 20, 22)], 
                     labels=c(1:16, 18, 20, 20, 22), 
                     expand=c(0,0)) +
  scale_colour_manual(values=mh_cols, guide="none") +
  theme_minimal() +
  theme(axis.title.x=element_blank(),
        axis.text.x=element_text(vjust = -1.5),
        panel.grid.minor=element_blank(),
        panel.grid.major.x=element_blank()) +
  labs("Example 'Manhattan' plot for HbA1c (analysis was not genome-wide)")
```

```{r pruning, eval=F}
prune_chromosome <- function(chr_df, pval_col, locus_width) {
  # Prune variants given a chromosome-specific summary statistic data frame
  df <- arrange(chr_df, !!sym(pval_col))  # Sort by ascending p-value
  pruned_df <- tibble()
  while(nrow(df) > 0) {
    pruned_df <- bind_rows(pruned_df, df[1, ])  # Add lowest p-value to pruned dataset
    df <- filter(df, (POS < df$POS[1] - locus_width / 2) |  # Remove rest of variants in that distance-based locus
                   (POS > df$POS[1] + locus_width / 2))
  }
  pruned_df
}

prune_variants <- function(ss_df, pval_col, thresh=5e-8, locus_width=1000000) {
  # Prune variants across all chromosomes using a simple distance-based approach
  if (min(ss_df[[pval_col]], na.rm=T) > thresh) return(tibble())
  ss_df %>%
    filter(!!sym(pval_col) < thresh) %>%
    mutate(POS = as.numeric(POS)) %>%
    nest(data=-CHR) %>%
    mutate(pruned_ss=map(data, prune_chromosome, pval_col, locus_width)) %>%
    unnest(pruned_ss) %>%
    select(-data) %>%
    dplyr::rename(index_var=SNP)
}

# pruned_res <- primary_res %>%
#   select(biomarker, gene, SNP=RSID, CHR, POS, gene, P_Value_Interaction) %>%
#   nest(-biomarker) %>%
#   mutate(pruned_bm_res = map(data, prune_variants, "P_Value_Interaction", thresh=0.05)) %>%
#   select(-data) %>%
#   unnest(pruned_bm_res)
```

```{r load-gene-sumstats-FUMA, eval=F}
read_gene_sumstats <- function(bm, model) {
  fn <- paste0("../data/processed/gwis/fuma/", bm, "_", model, "/magma.genes.out")
  read_tsv(fn) %>%
    select(ens_gene=GENE, gene=SYMBOL, z=ZSTAT, P) %>%
    mutate(q = p.adjust(P, method="BH"))
}

gene_res_list <- lapply(setNames(biomarkers, biomarkers), function(bm) {
  lapply(setNames(diet_vars, diet_vars), function(model) {
    read_gene_sumstats(bm, model)
  })
})
```
